man/pax11publish.1
.TH pax11publish 1 User Manuals
.SH NAME
pax11publish \- PulseAudio X11 Credential Utility
.SH SYNOPSIS
\fBpax11publish -h\fB

pax11publish [\fIoptions\fB] [-d\fB]

pax11publish [\fIoptions\fB] -e\fB

pax11publish [\fIoptions\fB] -i\fB

pax11publish [\fIoptions\fB] -r\fB
\f1
.SH DESCRIPTION
The \fIpax11publish\f1 utility can be used to dump or manipulate the PulseAudio server credentials that can be stored as properties on the X11 root window.

Please note that the loadable module \fImodule-x11-publish\f1 exports the same information directly from the PulseAudio sound server, and should in most cases be used in preference over this tool.

Use the following command to dump the raw PulseAudio-specific data that is stored in your X11 root window:

xprop -root | grep ^PULSE_
.SH OPTIONS
.TP
\fB-h\f1
Show help.
.TP
\fB-d\f1
Read the PulseAudio server credentials currently set on the X11 root window and dump them in a human readable form. This reads the PULSE_SERVER, PULSE_SINK, PULSE_SOURCE and PULSE_COOKIE properties.
.TP
\fB-i\f1
Similar to \fB-d\f1, however dumps them in a Bourne shell compatible format so they may be used together with the \fIeval\f1 shell command to set the $PULSE_SERVER, $PULSE_SINK, $PULSE_SOURCE environment variables. Also reads the authentication cookie from the root window and stores it in \fI~/.pulse-cookie\f1. 
.TP
\fB-e\f1
Export the currently locally used sound server, sink, source configuration to the X11 root window. This takes the data from the $PULSE_SERVER, $PULSE_SINK, $PULSE_SOURCE environment variables and combines them with the data from \fI~/.pulse/client.conf\f1 (or \fI/usr/local/etc/pulse/client.conf\f1 if that file does not exist). If specific options are passed on the command line (\fB-S\f1, \fB-O\f1, \fB-I\f1, \fB-c\f1, see below), they take precedence. Also uploads the local authentication cookie \fI~/.pulse-cookie\f1 to the X11 server.
.TP
\fB-r\f1
Removes the configured PulseAudio configuration from the X11 root window.
.TP
\fB-D\f1 \fIDISPLAY\f1
Connect to the specified X11 display, instead of the default one configured in $DISPLAY.
.TP
\fB-S\f1 \fISERVER\f1
Only valid for \fB-e\f1: export the specified PulseAudio server as default to the X11 display instead of the one configured via local configuration.
.TP
\fB-O\f1 \fISINK\f1
Only valid for \fB-e\f1: export the specified sink as default sink to the X11 display instead of the one configured via local configuration.
.TP
\fB-I\f1 \fISOURCE\f1
Only valid for \fB-e\f1: export the specified source as default to the X11 display instead of the one configured via local configuration.
.TP
\fB-c\f1 \fIFILE\f1
Only valid for \fB-e\f1: export the PulseAudio authentication cookie stored in the specified file to the X11 display instead of the one stored in \fI~/.pulse-cookie\f1.
.SH AUTHORS
The PulseAudio Developers <mzchyfrnhqvb (at) 0pointer (dot) net>; PulseAudio is available from \fBhttp://pulseaudio.org/\f1
.SH SEE ALSO
\fBpulseaudio(1)\f1, \fBxprop(1)\f1


ChangeLog

2009-04-14	Lennart Poettering
	ca9cd14: add a few missing initializations
	40d71e1: explcitly ignore return values of some functions marked with gcc's warn_unused_result attribute
	845b312: bump version for final 0.9.15 release
	a4cea4e: core: when applying delay memblockq take monitored sink latency into account
	2d94c93: pacat: add missing newline
	4ee4a55: core: use pa_{source
2009-04-13	Lennart Poettering
	16324fc: sort LINGUAS
	a654155: run make update-po
	c96f6f3: update LINGUAS
	0948dca: Merge commit 'origin/master-tx'
	fe8b10c: core: introduce new 'reference' volume for sinks
	49dcf09: alsa: include the alsa mixer control that is used in the property list
	6fd8fd1: alsa: store mixer controls to use in profile data
	89f74cb: alsa: when passing emptry mixer control name, force sw volume
	237a9e1: volume: increase dB range to -90dB
	0ac038e: client-conf-x11: unbreak autospawn due to stale X11 properties
	0aed5ea: client-conf: when  is set, disable autospawn setting
	4cc4cbd: client-conf: make setting a default server independant from the autospawn setting
	20aba71: proplist-util: use pa_session_id() instead of accessing 7b816367b01393ed3e3e650047d78f6e-1239640487.203609-1061245823 directly
	43650de: client-conf: modernize a few things
	a36197c: print session id when starting up
	1d8da03: core-util: filter utf8 in pa_machine_id()
	1b4e5f1: core-util: add pa_session_id()
	17f1784: cork-music-on-phone: make sure that we don't check the refcnt of pa_core when the daemon goes down
	62db10c: lirc: fix logic behind mute buttons
	66d2184: mmkbd: get rid of support for ancient kernels
	f1d3dfb: mmkbd,lirc: make use of pa_assert_not_reached()
	270a698: lirc, mmkbd: extend controllable volume range to PA_VOLUME_MAX
	6d218e9: api: introduce PA_VOLUME_MAX
	e9dd7a5: lirc: drop lirc_in_use, it's made redundant by PA_MODULE_LOAD_ONCE
	d8de5d3: make sure we never overflow when calculating sleep time

2009-04-13	warrink
	2ba882c: Sending translation for po/nl.po

2009-04-11	ifelix
	362ec7b: Sending translation for Tamil

2009-04-10	Lennart Poettering
	6eaeaea: Downgrade default log level to NOTICE to follow documentation
	a26c945: add missing languages to LINGUAS
	3e29fd7: prepare another snapshot
	991cb06: set fixed latencies at more places where appropriate

2009-04-06	Finn Thain
	80e18c8: make dbus optional during build

2009-04-10	Lennart Poettering
	e011230: run make update-po
	d6f019e: Merge commit 'origin/master-tx'
	5388b44: alsa: when printing warning about bogus data from alsa include snd_pcm_dump()
	ee6657a: bluetooth: when starting up HSP stream, send 2 packets first, only afterwards enter one-read-one-write logic
	d77b28c: bluetooth: rework timing logic, properly implement latency callbacks
	e9a4dec: bluetooth: be a bit more verbose if we exit due to bad poll() revents flag
	48cff5b: bluetooth: rename sco to hsp also for the user
	f7c229d: core: add a seperate fixed_latency field for sinks/sources with fixed latency
	9ae8ca2: core: memory leak, fix ref counting when moving streams
	dcd4a73: dbus: memory leak, actually free dbus wrapper
	d827ecd: dbus: drop pa_ prefix from static symbol
	f8ebe85: protocol-native: downgrade message if we receive pcm block for dead stream
	5b87196: protocol-native: print underrun message only once for each underrun
	3507d1e: socket-server: memory leak, free machine id after use
	669703d: dbus: memory leak, free pending calls
	9ba9883: dbus: memory leak, free server id after use
	9ee6a41: bluetooth: memory leak, actually free discovery struct itself
	f65b276: interpol-test: make it easier to test corking only optionally

2009-04-09	elsupergomez
	dac687d: Sending translation for Spanish

2009-04-09	mgiri
	d1cf6eb: Sending translation for Oriya
	6c5bbd3: Sending translation for Oriya

2009-04-08	mgiri
	8e1a9b7: Sending translation for Oriya

2009-04-08	kkrothap
	27784c4: Sending translation for Telugu

2009-04-08	ifelix
	0474762: Sending translation for po/ta.po

2009-04-08	runab
	bb7beeb: Sending translation for Bengali (India)

2009-04-08	amitakhya
	89f5da8: Sending translation for Assamese
	b445f79: Sending translation for po/as.po

2009-04-08	Lennart Poettering
	6c04a1c: bluetooth: make sure to set max_request
	c32c6c8: introduce relative_volume field in sink_input and make use of it on sink flat volume change
	14e89d4: when calculating volume from dB use ceil()
	4ff41ec: print smallest attenuation/sample

2009-04-07	kkrothap
	091a0f2: Sending translation for po/te.po

2009-04-07	kmilos
	f5eca5d: Sending translation for po/sr@latin.po
	7202d52: Sending translation for Serbian

2009-04-07	mgiri
	5b0d1bf: Sending translation for Oriya
	54c895c: Sending translation for po/or.po

2009-04-07	runab
	7154f06: Sending translation for po/bn_IN.po

2009-04-07	Lennart Poettering
	aacb11b: update documentation regarding stream timing a bit
	c523b16: after propagating a sink volume change to the sink inputs recalculate their soft volumes
	93e14d3: we need to make our multiplications with linear values
	02686cc: reduce number of conversions to/from linear volumes
	d612fbb: compare with doubles, not integer
	e356a03: If the sink volume is lowered to 0 and then increased again, make sure all stream volumes follow instead of staying at 0
	2c2713a: make use of SO_TIMESTAMP timestamp for accuracy and leave smoother paused until we have data
	f204c0f: mark null sink as support dynamic latency
	298bd0b: adjust max_rewind/max_request whenever the latency changes
	e976034: send the source latency based on the MTU size
	61b0776: add suspend_within_thread() callbacks to pa_sink_input/pa_source_output
	35a4a0b: enable debugging output based on if DEBUG_DATA macro is set

2009-04-07	swkothar
	ff2dc1b: Sending translation for Gujarati

2009-04-07	shanky
	682c101: Sending translation for po/kn.po

2009-04-06	Lennart Poettering
	886ddc3: make sure we don't apply sampling rate fixes that bring the sampling freq > PA_RATE_MAX

src/utils/pax11publish.c
/* $Id: pax11publish.c 1426 2007-02-13 15:35:19Z ossman $ */

/***
  This file is part of PulseAudio.

  Copyright 2004-2006 Lennart Poettering

  PulseAudio is free software; you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published
  by the Free Software Foundation; either version 2 of the License,
  or (at your option) any later version.

  PulseAudio is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with PulseAudio; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
  USA.
***/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdio.h>
#include <getopt.h>
#include <string.h>
#include <assert.h>

#include <X11/Xlib.h>
#include <X11/Xatom.h>

#include <pulse/util.h>

#include <pulsecore/core-util.h>
#include <pulsecore/log.h>
#include <pulsecore/authkey.h>
#include <pulsecore/native-common.h>
#include <pulsecore/x11prop.h>

#include "../pulse/client-conf.h"

int main(int argc, char *argv[]) {
    const char *dname = NULL, *sink = NULL, *source = NULL, *server = NULL, *cookie_file = PA_NATIVE_COOKIE_FILE;
    int c, ret = 1;
    Display *d = NULL;
    enum { DUMP, EXPORT, IMPORT, REMOVE } mode = DUMP;

    while ((c = getopt(argc, argv, "deiD:S:O:I:c:hr")) != -1) {
        switch (c) {
            case 'D' :
                dname = optarg;
                break;
            case 'h':
                printf("%s [-D display] [-S server] [-O sink] [-I source] [-c file]  [-d|-e|-i|-r]\n\n"
                       " -d    Show current PulseAudio data attached to X11 display (default)\n"
                       " -e    Export local PulseAudio data to X11 display\n"
                       " -i    Import PulseAudio data from X11 display to local environment variables and cookie file.\n"
                       " -r    Remove PulseAudio data from X11 display\n",
                       pa_path_get_filename(argv[0]));
                ret = 0;
                goto finish;
            case 'd':
                mode = DUMP;
                break;
            case 'e':
                mode = EXPORT;
                break;
            case 'i':
                mode = IMPORT;
                break;
            case 'r':
                mode = REMOVE;
                break;
            case 'c':
                cookie_file = optarg;
                break;
            case 'I':
                source = optarg;
                break;
            case 'O':
                sink = optarg;
                break;
            case 'S':
                server = optarg;
                break;
            default:
                fprintf(stderr, "Failed to parse command line.\n");
                goto finish;
        }
    }

    if (!(d = XOpenDisplay(dname))) {
        pa_log("XOpenDisplay() failed");
        goto finish;
    }

    switch (mode) {
        case DUMP: {
            char t[1024];
            if (pa_x11_get_prop(d, "PULSE_SERVER", t, sizeof(t)))
                printf("Server: %s\n", t);
            if (pa_x11_get_prop(d, "PULSE_SOURCE", t, sizeof(t)))
                printf("Source: %s\n", t);
            if (pa_x11_get_prop(d, "PULSE_SINK", t, sizeof(t)))
                printf("Sink: %s\n", t);
            if (pa_x11_get_prop(d, "PULSE_COOKIE", t, sizeof(t)))
                printf("Cookie: %s\n", t);

            break;
        }

        case IMPORT: {
            char t[1024];
            if (pa_x11_get_prop(d, "PULSE_SERVER", t, sizeof(t)))
                printf("PULSE_SERVER='%s'\nexport PULSE_SERVER\n", t);
            if (pa_x11_get_prop(d, "PULSE_SOURCE", t, sizeof(t)))
                printf("PULSE_SOURCE='%s'\nexport PULSE_SOURCE\n", t);
            if (pa_x11_get_prop(d, "PULSE_SINK", t, sizeof(t)))
                printf("PULSE_SINK='%s'\nexport PULSE_SINK\n", t);

            if (pa_x11_get_prop(d, "PULSE_COOKIE", t, sizeof(t))) {
                uint8_t cookie[PA_NATIVE_COOKIE_LENGTH];
                size_t l;
                if ((l = pa_parsehex(t, cookie, sizeof(cookie))) != sizeof(cookie)) {
                    fprintf(stderr, "Failed to parse cookie data\n");
                    goto finish;
                }

                if (pa_authkey_save(cookie_file, cookie, l) < 0) {
                    fprintf(stderr, "Failed to save cookie data\n");
                    goto finish;
                }
            }

            break;
        }

        case EXPORT: {
            pa_client_conf *conf = pa_client_conf_new();
            uint8_t cookie[PA_NATIVE_COOKIE_LENGTH];
            char hx[PA_NATIVE_COOKIE_LENGTH*2+1];
            assert(conf);

            if (pa_client_conf_load(conf, NULL) < 0) {
                fprintf(stderr, "Failed to load client configuration file.\n");
                goto finish;
            }

            if (pa_client_conf_env(conf) < 0) {
                fprintf(stderr, "Failed to read environment configuration data.\n");
                goto finish;
            }

            pa_x11_del_prop(d, "PULSE_SERVER");
            pa_x11_del_prop(d, "PULSE_SINK");
            pa_x11_del_prop(d, "PULSE_SOURCE");
            pa_x11_del_prop(d, "PULSE_ID");
            pa_x11_del_prop(d, "PULSE_COOKIE");

            if (server)
                pa_x11_set_prop(d, "PULSE_SERVER", server);
            else if (conf->default_server)
                pa_x11_set_prop(d, "PULSE_SERVER", conf->default_server);
            else {
                char hn[256];
                if (!pa_get_fqdn(hn, sizeof(hn))) {
                    fprintf(stderr, "Failed to get FQDN.\n");
                    goto finish;
                }

                pa_x11_set_prop(d, "PULSE_SERVER", hn);
            }

            if (sink)
                pa_x11_set_prop(d, "PULSE_SINK", sink);
            else if (conf->default_sink)
                pa_x11_set_prop(d, "PULSE_SINK", conf->default_sink);

            if (source)
                pa_x11_set_prop(d, "PULSE_SOURCE", source);
            if (conf->default_source)
                pa_x11_set_prop(d, "PULSE_SOURCE", conf->default_source);

            pa_client_conf_free(conf);

            if (pa_authkey_load_auto(cookie_file, cookie, sizeof(cookie)) < 0) {
                fprintf(stderr, "Failed to load cookie data\n");
                goto finish;
            }

            pa_x11_set_prop(d, "PULSE_COOKIE", pa_hexstr(cookie, sizeof(cookie), hx, sizeof(hx)));
            break;
        }

        case REMOVE:
            pa_x11_del_prop(d, "PULSE_SERVER");
            pa_x11_del_prop(d, "PULSE_SINK");
            pa_x11_del_prop(d, "PULSE_SOURCE");
            pa_x11_del_prop(d, "PULSE_ID");
            pa_x11_del_prop(d, "PULSE_COOKIE");
            break;

        default:
            fprintf(stderr, "No yet implemented.\n");
            goto finish;
    }

    ret = 0;

finish:

    if (d) {
        XSync(d, False);
        XCloseDisplay(d);
    }

    return ret;
}
